app/src/main/java/gifapp/jsm/myapp/app/interview/jsmtech/gifapp/MainActivity.java

```diff
@@ -0,0 +1,185 @@
+                    // othebe: No need to pass variables since they are part of object.
+                    makeGifJSONRequest(trendingUrl + trendingOffset);
+                }
+            }
+        });
+
+        // enabling search in real time
+        mSearch.addTextChangedListener(new TextWatcher() {
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+            }
+
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+                if (s.toString().length() != 0) { // something is entered in the EditText
+                    offset = 0; // every time something new is entered, the search reinitializes
+                    calledMethod = 1; // search method
+                    searchGifs(s.toString()); // calling the search method
+                } else {
+                    // shows trending GIFs by default
+                    trendingOffset = 0;
+                    calledMethod = 0;
+                    makeGifJSONRequest(trendingUrl + trendingOffset);
+                }
+            }
+
+            @Override
+            public void afterTextChanged(Editable s) {
+            }
+        });
+    }
+
+
+    // search method which in turn calls the volley method
+    public void searchGifs(String s) {
+        String searchString = s.trim().replaceAll("\\s+", "+");
+        searchUrl = "http://api.giphy.com/v1/gifs/search?q=" + searchString + "&api_key=" + API_KEY + "&offset=";
+        calledMethod = 1;
+        makeGifJSONRequest(searchUrl + 0);
+    }
+
+    // volley request for fetching the JSON data and adding it to the list of gifs
+    public void makeGifJSONRequest(String url) {
+        mProgressBar.setVisibility(View.VISIBLE);
+        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, url, null, new Response.Listener<JSONObject>() {
+            @Override
+            public void onResponse(JSONObject response) {
+                try {
+                    JSONArray j = response.getJSONArray("data");
+                    // delete all data when its a new call
+                    if ((offset == 0 && calledMethod == 1) || (trendingOffset == 0 && calledMethod == 0)) {
+                        listOfGifs.clear();
+                        gifAdapter.notifyDataSetChanged();
+                        mRecycleView.smoothScrollToPosition(0);
+                    }
+                    // fetching JSON response
+                    for (int i = 0; i < j.length(); i++) {
+                        // getting the GIF url for all the trending gifs
+                        String url = j.getJSONObject(i).getJSONObject("images").getJSONObject("fixed_height").getString("url");
+                        // adding to the list
+                        GIFDetails gifDetails = new GIFDetails();
+                        gifDetails.setGifUrl(url);
+                        listOfGifs.add(gifDetails);
```

Comment:
One option would be to move the responsibility of managing the items to the adapter, i.e defining add/remove methods on the adapter instead.